micro-rdk/src/common/ota.rs

```diff
@@ -0,0 +1,359 @@
+        grpc_client::H2Timer,
+    },
+    proto::app::v1::ServiceConfig,
+};
+
+#[cfg(feature = "esp32")]
+use crate::esp32::esp_idf_svc::{
+    ota::{EspFirmwareInfoLoader, EspOta},
+    sys::EspError,
+};
+use bincode::Decode;
+use futures_lite::AsyncWriteExt;
+use http_body_util::{BodyExt, Empty};
+use hyper::{body::Bytes, client::conn::http2, Request};
+use thiserror::Error;
+
+// TODO(RSDK-9200): set according to active partition scheme
+const OTA_MAX_IMAGE_SIZE: usize = 1024 * 1024 * 4; // 4MB
+pub const OTA_MODEL_TYPE: &str = "ota_service";
+pub const OTA_MODEL_TRIPLET: &str = "rdk:builtin:ota_service";
+
+/// https://github.com/espressif/esp-idf/blob/ce6085349f8d5a95fc857e28e2d73d73dd3629b5/components/esp_app_format/include/esp_app_desc.h#L42
+/// typedef struct {
+///     uint32_t magic_word;        /*!< Magic word ESP_APP_DESC_MAGIC_WORD */
+///     uint32_t secure_version;    /*!< Secure version */
+///     uint32_t reserv1[2];        /*!< reserv1 */
+///     char version[32];           /*!< Application version */
+///     char project_name[32];      /*!< Project name */
+///     char time[16];              /*!< Compile time */
+///     char date[16];              /*!< Compile date*/
+///     char idf_ver[32];           /*!< Version IDF */
+///     uint8_t app_elf_sha256[32]; /*!< sha256 of elf file */
+///     uint16_t min_efuse_blk_rev_full; /*!< Minimal eFuse block revision supported by image, in format: major * 100 + minor */
+///     uint16_t max_efuse_blk_rev_full; /*!< Maximal eFuse block revision supported by image, in format: major * 100 + minor */
+///     uint32_t reserv2[19];       /*!< reserv2 */
+/// } esp_app_desc_t;
+///
+/// should be 256 bytes in size
+#[repr(C)]
+#[derive(Decode, Debug, PartialEq)]
+struct EspAppDesc {
+    //TODO verify this is doing what I think it's doing, validate results
+    /// ESP_APP_DESC_MAGIC_WORD (0xABCD5432)
+    magic_word: u32,
+    secure_version: u32,
+    reserv1: [u32; 2],
+    /// application version
+    version: [u8; 32],
+    project_name: [u8; 32],
+    /// compile time
+    time: [u8; 16],
+    /// compile date
+    date: [u8; 16],
+    idf_ver: [u8; 32],
+    app_elf_sha256: [u8; 32],
+    /// minimal eFuse block revision supported by image, in format: major * 100 + minor
+    min_efuse_blk_rev_full: u16,
+    /// maximal eFuse block revision supported by image, in format: major * 100 + minor
+    max_efuse_blk_rev_full: u16,
+    reserv2: [u32; 19],
+}
+
+// TODO(RSDK-9214)
+#[derive(Error, Debug)]
+pub enum OtaError {
+    #[error("{0}")]
+    ConfigError(String),
+    #[cfg(feature = "esp32")]
+    #[error(transparent)]
+    EspError(#[from] EspError),
+    #[error("failed to initialize ota process")]
+    InitError,
+    #[error("new image size is invalid: {0} bytes")]
+    InvalidImageSize(usize),
+    #[error("{0}")]
+    Other(String),
+}
+
+#[cfg(feature = "esp32")]
+type OtaConnector = crate::esp32::tcp::Esp32H2Connector;
+#[cfg(not(feature = "esp32"))]
```

Comment:
maybe I forgot from the previous PR, but why are we allowing cases where we're not on an esp32? Isn't OTA meaningless otherwise?

Reply:
testing, specifically around hyper, config checks, and inspecting the binary to be downloaded

Reply:
That's fair for development purposes maybe, but isn't this confusing for merged-in code? It took me a while to see that OTA compiles for native but makes http calls that essentially do nothing.

Reply:
Not sure what you mean by nothing. The calls are for downloading the OTA binary and inspecting it. This would also be useful for  verifying future work we talked about for editing the binary signatures directly. 

Should I put it back and take it out between every PR for this feature? Or clean it up at the end if it doesn't prove useful beyond the planned work?

Reply:
I guess to that point, what is the purpose of having a native build?

Reply:
I mean the native build still tests the connection infrastructure abstraction and acts as a server. In native "OTA", there's no updating happening so it doesn't feel like a fair comparison. But we can leave it in and revisit at the end of the epic